ratis-server/src/main/java/org/apache/ratis/server/impl/ServerState.java

```diff
@@ -441,7 +441,9 @@ void installSnapshot(InstallSnapshotRequestProto request) throws IOException {
     StateMachine sm = server.getStateMachine();
     sm.pause(); // pause the SM to prepare for install snapshot
     snapshotManager.installSnapshot(request, sm);
-    updateInstalledSnapshotIndex(TermIndex.valueOf(request.getSnapshotChunk().getTermIndex()));
+    if (request.getSnapshotChunk().getDone()) {
+      updateInstalledSnapshotIndex(TermIndex.valueOf(request.getSnapshotChunk().getTermIndex()));
+    }
```

Comment:
@kaijchen , thanks a lot for working on this!  Let's combine `updateInstalledSnapshotIndex` with `reloadStateMachine`; see https://issues.apache.org/jira/secure/attachment/13056909/868_review.patch